### PR TITLE
Add FXIOS-7912 [v122] Display the same ad if bottom sheet is closed and reopened again

### PR DESCRIPTION
**Pull request by PARAIPAN9**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-7912)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/17664)

## :bulb: Description

- Cache the ads if the review checker is closed and reopened for the same productId.
- Clear cache if the URL has changed.
## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/Fakespot/ProductAdsCacheTests.swift [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            cache.cacheAds(ads1, forKey: key)
+
+            let cachedAds = cache.getCachedAds(forKey: key)
+            XCTAssertNotNil(cachedAds, "Cached ads should not be nil")
+            XCTAssertEqual(cachedAds, ads1, "Cached ads should match the original ads")
+
+            // Test getting ads for a non-existent key
+            let nonExistentAds = cache.getCachedAds(forKey: "nonExistentKey")
+            XCTAssertNil(nonExistentAds, "Cached ads for a non-existent key should be nil")
+        }
```
Comment (razvanlitianu) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/Fakespot/ProductAdsCacheTests.swift:80]: nit: The indentations seems a bit off.

File: Client/Frontend/Fakespot/Client/Response/ProductAdsCache.swift [hunk 2]
```diff
@@ -0,0 +1,33 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+
+class ProductAdsCache {
+    static let shared = ProductAdsCache()
+    private let cache = NSCache<NSString, CachedAds>()
```
Comment (razvanlitianu) [inline Client/Frontend/Fakespot/Client/Response/ProductAdsCache.swift:9]: Just curious why we don't use a dictionary, do we persist the caches?

File: Client/Frontend/Toolbar+URLBar/TabLocationView.swift [hunk 3]
```diff
@@ -48,6 +48,7 @@ class TabLocationView: UIView, FeatureFlaggable {
     private let menuBadge = BadgeWithBackdrop(imageName: ImageIdentifiers.menuBadge, backdropCircleSize: 32)
 
     var url: URL? {
+        willSet { handleShoppingAdsCacheURLChange(newURL: newValue) }
```
Comment (razvanlitianu) [inline Client/Frontend/Toolbar+URLBar/TabLocationView.swift:51]: Is this the source of truth from where we know the url changed?

Comment (PARAIPAN9) [reply Client/Frontend/Toolbar+URLBar/TabLocationView.swift:51]: It is always in sync whenever the url changes.

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/Fakespot/FakespotViewModelTests.swift [hunk 4]
```diff
@@ -8,30 +8,113 @@ import XCTest
...
+        ]
 
         override func fetchProductAnalysisData(maxRetries: Int = 3, retryTimeout: Int = 100) async throws -> ProductAnalysisResponse? {
             if shouldThrowError {
                 throw NSError(domain: "MockErrorDomain", code: 123, userInfo: nil)
             }
             return .empty
         }
+
+        override func fetchProductAdsData() async -> [ProductAdsResponse] {
```
Comment (razvanlitianu) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/Fakespot/FakespotViewModelTests.swift:56]: This boolean is not really concludent way of telling us if we should fetch ads or not. Ideally we would mock prefs and the cache and not override this method.

Comment (PARAIPAN9) [reply firefox-ios/firefox-ios-tests/Tests/ClientTests/Fakespot/FakespotViewModelTests.swift:56]: Perhaps there's confusion about the naming. I aim to simulate server responses, both with data and without data, by overriding the method.

File: Client/Frontend/Fakespot/Client/Response/ProductAdsCache.swift [hunk 5]
```diff
@@ -0,0 +1,33 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+
+class ProductAdsCache {
```
Comment (OrlaM) [inline Client/Frontend/Fakespot/Client/Response/ProductAdsCache.swift:7]: Would it be worth making this an actor instead of a class for thread safety?

**Review by OrlaM — APPROVED** [review]

**Review by yoanarios — APPROVED** [review]

File: Client/Frontend/Toolbar+URLBar/TabLocationView.swift [hunk 6]
```diff
@@ -326,6 +327,15 @@ class TabLocationView: UIView, FeatureFlaggable {
                                      object: .shoppingProductPageVisits)
     }
 
+    private func handleShoppingAdsCacheURLChange(newURL: URL?) {
+        if url?.displayURL != newURL,
```
Comment (thatswinnie) [inline Client/Frontend/Toolbar+URLBar/TabLocationView.swift:331]: I suggest converting this to a `guard` for better readability.

File: Client/Frontend/Fakespot/FakespotViewModel.swift [hunk 7]
```diff
@@ -400,6 +397,25 @@ class FakespotViewModel {
         }
     }
 
+    func loadProductAds(for productId: String?) async -> [ProductAdsResponse] {
+        if let productId,
```
[the remark this answers is not in the document]
Comment (PARAIPAN9) [reply Client/Frontend/Fakespot/FakespotViewModel.swift:401]: The ads are not shown even though they are cached. This guard statement ```guard viewModel.areAdsEnabled else { return nil }``` in ```FakespotViewController``` checks whether ads are enabled and determines if the view needs to be shown or not.

**Review by thatswinnie — APPROVED** [review]

**Pull request merged**